core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala

```diff
@@ -0,0 +1,369 @@
+   */
+  @Test
+  def testAppendReplicaFetchWithSchedulerCheckForShrinkIsr(): Unit = {
+    val active = new AtomicBoolean(true)
+
+    val future = scheduleShrinkIsr(active, mockTimeSleepMs = 0)
+    concurrentProduceFetch(numProducers, numReplicaFetchers, numRecordsPerProducer, appendDelayMs)
+    active.set(false)
+    future.get(30, TimeUnit.SECONDS)
+
+    verifyMetrics(expectLockContention = false)
+    assertEquals(0, writeLockCount.get())
+    assertTrue(readLockCount.get() > 0)
+  }
+
+  /**
+   * Verifies concurrent produce and replica fetch log read result update with ISR updates. This
+   * can result in delays in processing produce and replica fetch requets since write lock is obtained,
+   * but it should complete without any failures.
+   */
+  @Test
+  def testAppendReplicaFetchWithUpdateIsr(): Unit = {
+    val active = new AtomicBoolean(true)
+
+    val future = scheduleShrinkIsr(active, mockTimeSleepMs = 10000)
+    concurrentProduceFetch(numProducers, numReplicaFetchers, numRecordsPerProducer, appendDelayMs)
+    active.set(false)
+    future.get(30, TimeUnit.SECONDS)
+
+    verifyMetrics(expectLockContention = true)
+    assertTrue(writeLockCount.get() > 0)
+    assertTrue(readLockCount.get() > 0)
+  }
+
+  private def verifyMetrics(expectLockContention : Boolean): Unit = {
+    def metricValue(metricName: MetricName): Double = {
+      val nanos = metrics.metric(metricName).metricValue.asInstanceOf[Double]
+      nanos / TimeUnit.MILLISECONDS.toNanos(1)
+    }
+
+    val appendAvg = metricValue(appendLatencyAvg)
+    val appendMax = metricValue(appendLatencyMax)
+    val updateFollowerFetchStateAvg = metricValue(updateFollowerFetchStateLatencyAvg)
+    val updateFollowerFetchStateMax = metricValue(updateFollowerFetchStateLatencyMax)
+    val metricsValues = s"appendAvg=$appendAvg appendMax=$appendMax updateFollowerFetchStateAvg=$updateFollowerFetchStateAvg updateFollowerFetchStateMax=$updateFollowerFetchStateMax"
+    info(s"Metrics: $metricsValues")
+    val errorMessage = s"Unexpected metrics: $metricsValues"
+
+    val toleranceMs = if (expectLockContention) (appendDelayMs + isrUpdateDelayMs) * 2 else 200
+    assertTrue(errorMessage, appendAvg < appendDelayMs + toleranceMs)
+    assertTrue(errorMessage, appendMax < appendDelayMs + toleranceMs + 500)
+    assertTrue(errorMessage, updateFollowerFetchStateAvg <= isrCheckDelayMs + toleranceMs)
+    assertTrue(errorMessage, updateFollowerFetchStateMax <= isrCheckDelayMs + toleranceMs + 500)
+  }
+
+  private def concurrentProduceFetch(numProducers: Int, numReplicaFetchers: Int, numRecords: Int, appendDelayMs: Long): Unit = {
+    val followerQueues = (0 until numReplicaFetchers).map(_ => new ArrayBlockingQueue[MemoryRecords](2))
+
+    val totalRecords = numRecordsPerProducer * numProducers
+    val futures = new ArrayBuffer[Future[_]]()
+
+    (0 until numProducers).foreach(_ => {
+      futures += executorService.submit(new Runnable {
+        override def run(): Unit = {
```

Comment:
I think now we can use the native scala lambdas. Instead of `new Runnable`, we can do `() =>`